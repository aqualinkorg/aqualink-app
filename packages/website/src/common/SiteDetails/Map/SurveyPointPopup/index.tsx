import React from "react";
import {
  Card,
  Grid,
  Typography,
  withStyles,
  WithStyles,
  createStyles,
} from "@material-ui/core";
import { Popup } from "react-leaflet";

import { Pois } from "../../../../store/Reefs/types";
import Link from "../../../Link";

const SurveyPointPopup = ({
  reefId,
  point,
  classes,
}: SurveyPointPopupProps) => {
  return (
    <Popup closeButton={false} autoPan={false}>
      <Card className={classes.surveyPointPopup}>
        <Grid
          container
          alignItems="center"
          justify="space-between"
          item
          spacing={2}
        >
          <Grid title={point.name || ""} item className={classes.nameWrapper}>
            <Typography
              className={classes.name}
              variant="h6"
              color="textSecondary"
            >
              {point.name}
            </Typography>
          </Grid>
          <Grid item>
            <Link
              to={`/reefs/${reefId}/points/${point.id}`}
              isIcon
              tooltipTitle="View survey point"
            />
          </Grid>
        </Grid>
      </Card>
    </Popup>
  );
};

const styles = () =>
  createStyles({
    surveyPointPopup: {
      minWidth: 150,
      maxWidth: 240,
      minHeight: 50,
      padding: 16,
    },
    nameWrapper: {
      maxWidth: "80%",
    },
    name: {
      overflowWrap: "break-word",
    },
  });

interface SurveyPointPopupIncomingProps {
  reefId: number;
  point: Pois;
}

type SurveyPointPopupProps = SurveyPointPopupIncomingProps &
  WithStyles<typeof styles>;

export default withStyles(styles)(SurveyPointPopup);
