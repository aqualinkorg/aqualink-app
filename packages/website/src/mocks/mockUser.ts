import { User } from "../store/User/types";

export const mockUser: User = {
  adminLevel: "default",
  administeredReefs: [],
  email: "mail@mail.com",
  firebaseUid: "RaNdOm StRiNg",
  fullName: "Mock User",
  organization: "Some organization",
  token: "RaNdOm StRiNg",
};
