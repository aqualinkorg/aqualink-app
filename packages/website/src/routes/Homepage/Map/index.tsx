import React, { useState, useRef, useEffect } from "react";
import { useSelector, useDispatch } from "react-redux";
import { Map, TileLayer, Marker, Popup as LeafletPopup } from "react-leaflet";
import L from "leaflet";
import {
  createStyles,
  withStyles,
  WithStyles,
  CircularProgress,
} from "@material-ui/core";

import {
  reefsListSelector,
  reefsListLoadingSelector,
} from "../../../store/Reefs/reefsListSlice";
import {
  reefOnMapSelector,
  unsetReefOnMap,
} from "../../../store/Homepage/homepageSlice";
import { Reef } from "../../../store/Reefs/types";
import Popup from "./Popup";
import { coloredBuoy } from "./utils";
import {
  colorFinder,
  degreeHeatingWeeksCalculator,
} from "../../../helpers/degreeHeatingWeeks";

const HomepageMap = ({ classes }: HomepageMapProps) => {
  const mapRef = useRef<Map>(null);
  const markerRef = useRef<Marker>(null);
  const dispatch = useDispatch();
  const reefOnMap = useSelector(reefOnMapSelector);
  const reefsList = useSelector(reefsListSelector);
  const loading = useSelector(reefsListLoadingSelector);
  const [center, setCenter] = useState<[number, number]>([0, 0]);
  const [zoom, setZoom] = useState<number>(2);

  useEffect(() => {
    const { current } = mapRef;
    if (current && current.leafletElement) {
      const map = current.leafletElement;
      const southWest = L.latLng(-90, -240);
      const northEast = L.latLng(90, 240);
      const bounds = L.latLngBounds(southWest, northEast);
      map.setMaxBounds(bounds);
      map.flyTo(center, zoom, { duration: 1 });
    }
  }, [center, zoom, mapRef]);

  // Add Sofar Tiles
  useEffect(() => {
    const { current } = mapRef;
    if (current && current.leafletElement) {
      const map = current.leafletElement;

      const layerControl = L.control.layers(undefined).addTo(map);
      const sofarLayers = [
        {
          name: "Sea Surface Temperature",
          model: "HYCOM",
          variableID: "seaSurfaceTemperature",
          cmap: "turbo",
        },
        {
          name: "NOAA Degree Heating Week",
          model: "NOAACoralReefWatch",
          variableID: "degreeHeatingWeek",
          cmap: "noaacoral",
        },
      ];

      const sofarUrl = "https://api.sofarocean.com/marine-weather/v1/models/";
      const { REACT_APP_SOFAR_API_TOKEN: token } = process.env;
      sofarLayers.forEach((layer) => {
        layerControl.addOverlay(
          L.tileLayer(
            `${sofarUrl}${layer.model}/tile/{z}/{x}/{y}.png?colormap=${layer.cmap}&token=${token}&variableID=${layer.variableID}`,
            { opacity: 0.5 }
          ),
          layer.name
        );
      });
    }
  }, [loading]);

  useEffect(() => {
    const markerCurrent = markerRef.current;
    const mapCurrent = mapRef.current;
    if (
      markerCurrent &&
      markerCurrent.leafletElement &&
      mapCurrent &&
      mapCurrent.leafletElement &&
      reefOnMap?.polygon?.type === "Point"
    ) {
      const map = mapCurrent.leafletElement;
      const marker = markerCurrent.leafletElement;
      map.flyTo(
        [reefOnMap.polygon.coordinates[1], reefOnMap.polygon.coordinates[0]],
        6,
        { duration: 1 }
      );
      marker.openPopup();
    }
  }, [reefOnMap, markerRef, mapRef]);

  return (
    <>
      {loading ? (
        <div className={classes.loading}>
          <CircularProgress size="10rem" thickness={1} />
        </div>
      ) : (
        <Map
          maxBoundsViscosity={1.0}
          className={classes.map}
          ref={mapRef}
          center={center}
          zoom={zoom}
          minZoom={2}
        >
          <TileLayer url="https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}" />
          {reefOnMap && reefOnMap.polygon.type === "Point" && (
            <Marker
              ref={markerRef}
              icon={L.divIcon({
                iconSize: [24, 24],
                iconAnchor: [8, 48],
                popupAnchor: [3, -48],
                html: `${coloredBuoy(
                  colorFinder(
                    degreeHeatingWeeksCalculator(
                      reefOnMap.latestDailyData.degreeHeatingDays
                    )
                  )
                )}`,
                className: "marker-icon",
              })}
              position={[
                reefOnMap.polygon.coordinates[1],
                reefOnMap.polygon.coordinates[0],
              ]}
            >
              <LeafletPopup closeButton={false} className={classes.popup}>
                <Popup reef={reefOnMap} />
              </LeafletPopup>
            </Marker>
          )}
          {reefsList.length > 0 &&
            reefsList.map((reef: Reef) => {
              if (reef.polygon.type === "Point") {
                const [lng, lat] = reef.polygon.coordinates;
                const { degreeHeatingDays } = reef.latestDailyData || {};
                const color = colorFinder(
                  degreeHeatingWeeksCalculator(degreeHeatingDays)
                );
                return (
                  <Marker
                    onClick={() => {
                      setZoom(6);
                      setCenter([lat, lng]);
                      dispatch(unsetReefOnMap());
                    }}
                    key={reef.id}
                    icon={L.divIcon({
                      iconSize: [24, 24],
                      iconAnchor: [8, 48],
                      popupAnchor: [3, -48],
                      html: `${coloredBuoy(color)}`,
                      className: "marker-icon",
                    })}
                    position={[
                      reef.polygon.coordinates[1],
                      reef.polygon.coordinates[0],
                    ]}
                  >
                    <LeafletPopup closeButton={false} className={classes.popup}>
                      <Popup reef={reef} />
                    </LeafletPopup>
                  </Marker>
                );
              }
              return null;
            })}
        </Map>
      )}
    </>
  );
};

const styles = () =>
  createStyles({
    map: {
      height: "100%",
      width: "100%",
    },
    popup: {
      width: "12vw",
    },
    loading: {
      height: "100%",
      width: "100%",
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
    },
  });

type HomepageMapProps = WithStyles<typeof styles>;

export default withStyles(styles)(HomepageMap);
