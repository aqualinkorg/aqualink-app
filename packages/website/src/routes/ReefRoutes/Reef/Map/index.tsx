import React, { useRef, useEffect, useCallback } from "react";
import { Map, TileLayer, Polygon, Marker } from "react-leaflet";
import { useDispatch, useSelector } from "react-redux";
import L from "leaflet";
import { withStyles, WithStyles, createStyles } from "@material-ui/core";

import { Reef, Position } from "../../../../store/Reefs/types";
import { mapBounds } from "../../../../helpers/mapBounds";

import marker from "../../../../assets/marker.png";
import {
  reefDraftSelector,
  setReefDraft,
} from "../../../../store/Reefs/selectedReefSlice";

const pinIcon = L.icon({
  iconUrl: marker,
  iconSize: [20, 30],
  iconAnchor: [10, 30],
  popupAnchor: [0, -41],
});

const ReefMap = ({ polygon, classes }: ReefMapProps) => {
  const dispatch = useDispatch();
  const mapRef = useRef<Map>(null);
  const markerRef = useRef<Marker>(null);
  const draftReef = useSelector(reefDraftSelector);

  const reverseCoords = (coordArray: Position[]): [Position[]] => {
    return [coordArray.map((coords) => [coords[1], coords[0]])];
  };

  useEffect(() => {
    const { current } = mapRef;
    if (current && current.leafletElement) {
      const map = current.leafletElement;
      // Initialize map's position to fit the given polygon
      if (polygon.type === "Polygon") {
        map.fitBounds(mapBounds(polygon));
      } else if (draftReef?.coordinates) {
        map.panTo(
          new L.LatLng(
            draftReef.coordinates.latitude,
            draftReef.coordinates.longitude
          )
        );
      } else {
        map.panTo(new L.LatLng(polygon.coordinates[1], polygon.coordinates[0]));
      }
    }
  }, [mapRef, draftReef, polygon]);

  const handleDragChange = useCallback(() => {
    const { current } = markerRef;
    if (current && current.leafletElement) {
      const mapMarker = current.leafletElement;
      const { lat, lng } = mapMarker.getLatLng();
      dispatch(
        setReefDraft({
          coordinates: {
            latitude: lat,
            longitude: lng,
          },
        })
      );
    }
  }, [dispatch]);

  return (
    <Map
      ref={mapRef}
      zoom={12}
      dragging
      scrollWheelZoom={false}
      className={classes.map}
    >
      <TileLayer url="https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}" />
      {polygon.type === "Polygon" ? (
        <Polygon positions={reverseCoords(...polygon.coordinates)} />
      ) : (
        <Marker
          ref={markerRef}
          draggable={Boolean(draftReef)}
          ondragend={handleDragChange}
          icon={pinIcon}
          position={[
            draftReef?.coordinates?.latitude || polygon.coordinates[1],
            draftReef?.coordinates?.longitude || polygon.coordinates[0],
          ]}
        />
      )}
    </Map>
  );
};

const styles = () => {
  return createStyles({
    map: {
      height: "100%",
      width: "100%",
      borderRadius: 4,
    },
  });
};

interface ReefMapIncomingProps {
  polygon: Reef["polygon"];
}

type ReefMapProps = WithStyles<typeof styles> & ReefMapIncomingProps;

export default withStyles(styles)(ReefMap);
