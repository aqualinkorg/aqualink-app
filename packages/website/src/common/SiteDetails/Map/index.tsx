import React, { useRef, useEffect, useCallback, useState } from "react";
import { Map, TileLayer, Polygon, Marker } from "react-leaflet";
import { useDispatch, useSelector } from "react-redux";
import L, { LatLngTuple } from "leaflet";
import "./plugins/leaflet-tilelayer-subpixel-fix";
import { withStyles, WithStyles, createStyles } from "@material-ui/core";
import { some } from "lodash";

import SurveyPointPopup from "./SurveyPointPopup";
import {
  Reef,
  Position,
  SpotterPosition,
  Pois,
  Point,
} from "../../../store/Reefs/types";
import { samePosition } from "../../../helpers/map";

import marker from "../../../assets/marker.png";
import buoy from "../../../assets/buoy-marker.svg";
import {
  reefDraftSelector,
  setReefDraft,
} from "../../../store/Reefs/selectedReefSlice";
import { userInfoSelector } from "../../../store/User/userSlice";
import { isManager } from "../../../helpers/user";
import pointIcon from "../../../assets/alerts/pin_nostress@2x.png";
import selectedPointIcon from "../../../assets/alerts/pin_warning@2x.png";

const pinIcon = L.icon({
  iconUrl: marker,
  iconSize: [20, 30],
  iconAnchor: [10, 30],
  popupAnchor: [0, -41],
});

const buoyIcon = L.icon({
  iconUrl: buoy,
  iconSize: [30, 40],
  iconAnchor: [10, 30],
  popupAnchor: [0, -41],
});

const surveyPointIcon = (selected: boolean) =>
  L.icon({
    iconUrl: selected ? selectedPointIcon : pointIcon,
    iconSize: [24, 27],
    iconAnchor: [12, 27],
    popupAnchor: [0, -27],
  });

const ReefMap = ({
  reefId,
  spotterPosition,
  polygon,
  surveyPoints,
  selectedPointId,
  surveyPointEditModeEnabled,
  editPointLatitude,
  editPointLongitude,
  onEditPointCoordinatesChange,
  classes,
}: ReefMapProps) => {
  const dispatch = useDispatch();
  const mapRef = useRef<Map>(null);
  const markerRef = useRef<Marker>(null);
  const editPointMarkerRer = useRef<Marker>(null);
  const draftReef = useSelector(reefDraftSelector);
  const user = useSelector(userInfoSelector);
  const [focusedPoint, setFocusedPoint] = useState<Pois>();

  const reverseCoords = (coordArray: Position[]): [Position[]] => {
    return [coordArray.map((coords) => [coords[1], coords[0]])];
  };

  const selectedSurveyPoint = surveyPoints.find(
    (item) => item.id === selectedPointId
  );

  const setCenter = (
    inputMap: L.Map,
    latLng: [number, number],
    zoom: number
  ) => {
    const maxZoom = Math.max(inputMap.getZoom() || 15, zoom);
    const pointBounds = L.latLngBounds(latLng, latLng);
    inputMap.flyToBounds(pointBounds, {
      maxZoom,
      duration: 2,
      paddingTopLeft: L.point(0, 200),
    });
  };

  // Fit the polygon constructed by the reef's center and its survey points
  const fitSurveyPointsPolygon = useCallback(
    (inputMap: L.Map, reefCenter: Point) => {
      inputMap.fitBounds(
        L.polygon([
          [reefCenter.coordinates[1], reefCenter.coordinates[0]],
          ...surveyPoints
            .filter((item) => item.polygon?.type === "Point")
            .map((item) => {
              const coords = item.polygon?.coordinates as Position;
              // Reverse coordinates since they come as [lng, lat]
              return [coords[1], coords[0]] as LatLngTuple;
            }),
        ]).getBounds()
      );
    },
    [surveyPoints]
  );

  useEffect(() => {
    if (
      mapRef?.current?.leafletElement &&
      focusedPoint?.polygon?.type === "Point"
    ) {
      const [lng, lat] = focusedPoint.polygon.coordinates;
      setCenter(mapRef.current.leafletElement, [lat, lng], 15);
    }
  }, [focusedPoint]);

  useEffect(() => {
    const { current } = mapRef;
    if (current?.leafletElement) {
      const map = current.leafletElement;
      // Initialize map's position to fit the given polygon
      if (polygon.type === "Polygon") {
        map.fitBounds(L.polygon(polygon.coordinates).getBounds());
      } else if (draftReef?.coordinates) {
        map.panTo(
          new L.LatLng(
            draftReef.coordinates.latitude || polygon.coordinates[1],
            draftReef.coordinates.longitude || polygon.coordinates[0]
          )
        );
      } else if (some(surveyPoints, (item) => item.polygon?.type === "Point")) {
        fitSurveyPointsPolygon(map, polygon);
      } else {
        map.panTo(new L.LatLng(polygon.coordinates[1], polygon.coordinates[0]));
      }
    }
  }, [draftReef, fitSurveyPointsPolygon, polygon, surveyPoints]);

  const handleDragChange = () => {
    const { current } = markerRef;
    if (current?.leafletElement) {
      const mapMarker = current.leafletElement;
      const { lat, lng } = mapMarker.getLatLng().wrap();
      dispatch(
        setReefDraft({
          coordinates: {
            latitude: lat,
            longitude: lng,
          },
        })
      );
    }
  };

  const handleEditPointDragChange = () => {
    const { current } = editPointMarkerRer;
    if (current && current.leafletElement && onEditPointCoordinatesChange) {
      const mapMarker = current.leafletElement;
      const { lat, lng } = mapMarker.getLatLng().wrap();
      onEditPointCoordinatesChange(lat.toString(), lng.toString());
    }
  };

  return (
    <Map
      ref={mapRef}
      maxZoom={17}
      zoom={13}
      dragging
      scrollWheelZoom={false}
      className={classes.map}
      tap={false}
    >
      <TileLayer url="https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}" />
      {polygon.type === "Polygon" ? (
        <Polygon positions={reverseCoords(...polygon.coordinates)} />
      ) : (
        <>
          {/* Marker to listen to survey point drag changes on edit mode */}
          {surveyPointEditModeEnabled && (
            <Marker
              ref={editPointMarkerRer}
              draggable={surveyPointEditModeEnabled}
              ondragend={handleEditPointDragChange}
              icon={surveyPointIcon(true)}
              zIndexOffset={100}
              position={[
                editPointLatitude ||
                  (selectedSurveyPoint?.polygon?.type === "Point" &&
                    selectedSurveyPoint.polygon.coordinates[1]) ||
                  polygon.coordinates[1],
                editPointLongitude ||
                  (selectedSurveyPoint?.polygon?.type === "Point" &&
                    selectedSurveyPoint.polygon.coordinates[0]) ||
                  polygon.coordinates[0],
              ]}
            />
          )}
          <Marker
            ref={markerRef}
            draggable={Boolean(draftReef)}
            ondragend={handleDragChange}
            icon={pinIcon}
            position={[
              draftReef?.coordinates?.latitude || polygon.coordinates[1],
              draftReef?.coordinates?.longitude || polygon.coordinates[0],
            ]}
          />
          {surveyPoints.map(
            (point) =>
              point?.polygon?.type === "Point" &&
              !samePosition(polygon, point.polygon) &&
              (point.id !== selectedPointId || !surveyPointEditModeEnabled) && ( // Hide selected survey point marker if it is in edit mode
                <Marker
                  key={point.id}
                  icon={surveyPointIcon(point.id === selectedPointId)}
                  position={[
                    point.polygon.coordinates[1],
                    point.polygon.coordinates[0],
                  ]}
                  onclick={() => setFocusedPoint(point)}
                >
                  <SurveyPointPopup reefId={reefId} point={point} />
                </Marker>
              )
          )}
        </>
      )}
      {!draftReef && spotterPosition && isManager(user) && (
        <Marker
          icon={buoyIcon}
          position={[
            spotterPosition.latitude.value,
            spotterPosition.longitude.value,
          ]}
        />
      )}
    </Map>
  );
};

const styles = () => {
  return createStyles({
    map: {
      height: "100%",
      width: "100%",
      borderRadius: 4,
    },
  });
};

interface ReefMapIncomingProps {
  reefId: number;
  polygon: Reef["polygon"];
  spotterPosition?: SpotterPosition | null;
  surveyPoints: Pois[];
  selectedPointId?: number;
  surveyPointEditModeEnabled?: boolean;
  editPointLatitude?: number | null;
  editPointLongitude?: number | null;
  onEditPointCoordinatesChange?: (lat: string, lng: string) => void;
}

ReefMap.defaultProps = {
  spotterPosition: null,
  selectedPointId: 0,
  surveyPointEditModeEnabled: false,
  editPointLatitude: null,
  editPointLongitude: null,
  onEditPointCoordinatesChange: () => {},
};

type ReefMapProps = WithStyles<typeof styles> & ReefMapIncomingProps;

export default withStyles(styles)(ReefMap);
