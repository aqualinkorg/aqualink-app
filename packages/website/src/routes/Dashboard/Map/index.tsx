import React from "react";
import {
  withStyles,
  WithStyles,
  createStyles,
  Box,
  Theme,
  useTheme,
  useMediaQuery,
} from "@material-ui/core";
import { LatLng } from "leaflet";

import Map from "../../HomeMap/Map";
import { getCollectionCenterAndBounds } from "../../../helpers/map";
import { CollectionDetails } from "../../../store/Collection/types";

const DashboardMap = ({ collection, classes }: DashboardMapProps) => {
  const theme = useTheme();
  const isMobile = useMediaQuery(theme.breakpoints.down("xs"));
  const [collectionCenter, collectionBounds] = getCollectionCenterAndBounds(
    collection
  );

  return (
    <Box className={classes.root}>
      <Map
        classes={{ map: classes.map }}
        initialCenter={collectionCenter || new LatLng(0, 0)}
        initialBounds={collectionBounds}
        initialZoom={3}
        collection={collection}
        showAlertLevelLegend={false}
        showWaterMark={false}
        geolocationEnabled={false}
        defaultLayerName="Heat Stress"
        legendBottom={isMobile ? 35 : 20}
        legendLeft={isMobile ? 2 : 4}
      />
    </Box>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    root: {
      borderRadius: 5,
      height: 480,
      marginTop: 46,
      [theme.breakpoints.only("md")]: {
        height: 420,
      },
      [theme.breakpoints.down("xs")]: {
        height: 450,
      },
    },
    map: {
      height: "100%",
      width: "100%",
      borderRadius: 5,
    },
  });

interface DashboardMapIncomingProps {
  collection: CollectionDetails;
}

type DashboardMapProps = DashboardMapIncomingProps & WithStyles<typeof styles>;

export default withStyles(styles)(DashboardMap);
