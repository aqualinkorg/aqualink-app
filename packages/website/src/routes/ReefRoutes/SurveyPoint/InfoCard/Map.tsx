import React from "react";
import {
  Grid,
  withStyles,
  WithStyles,
  createStyles,
  Theme,
} from "@material-ui/core";

import { Reef } from "../../../../store/Reefs/types";
import Map from "../../../../common/SiteDetails/Map";
import { FormField } from "../../../../hooks/useFormField";

const SurveyPointMap = ({
  reef,
  selectedPointId,
  editModeEnabled,
  editPointLatitude,
  editPointLongitude,
  onEditPointCoordinatesChange,
  classes,
}: SurveyPointMapProps) => {
  return (
    <Grid className={classes.mapWrapper} item xs={12} md={4}>
      <Map
        reefId={reef.id}
        polygon={reef.polygon}
        surveyPoints={reef.surveyPoints}
        selectedPointId={selectedPointId}
        surveyPointEditModeEnabled={editModeEnabled}
        editPointLatitude={parseFloat(editPointLatitude.value)}
        editPointLongitude={parseFloat(editPointLongitude.value)}
        onEditPointCoordinatesChange={onEditPointCoordinatesChange}
      />
    </Grid>
  );
};

const styles = (theme: Theme) =>
  createStyles({
    mapWrapper: {
      padding: 16,
      height: 280,
      [theme.breakpoints.down("sm")]: {
        height: 300,
      },
    },
  });

interface SurveyPointMapIncomingProps {
  reef: Reef;
  selectedPointId: number;
  editModeEnabled: boolean;
  editPointLatitude: FormField;
  editPointLongitude: FormField;
  onEditPointCoordinatesChange: (lat: string, lng: string) => void;
}

type SurveyPointMapProps = SurveyPointMapIncomingProps &
  WithStyles<typeof styles>;

export default withStyles(styles)(SurveyPointMap);
