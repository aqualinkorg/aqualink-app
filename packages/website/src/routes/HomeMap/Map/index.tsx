import React, { useEffect, useRef, useState } from "react";
import { useSelector } from "react-redux";
import { Map, TileLayer, Marker, Circle } from "react-leaflet";
import L, { LatLng, LayersControlEvent } from "leaflet";
import {
  createStyles,
  withStyles,
  WithStyles,
  CircularProgress,
  IconButton,
  Snackbar,
} from "@material-ui/core";
import { Alert } from "@material-ui/lab";
import MyLocationIcon from "@material-ui/icons/MyLocation";

import { reefsListLoadingSelector } from "../../../store/Reefs/reefsListSlice";
import { ReefMarkers } from "./Markers";
import { SofarLayers } from "./sofarLayers";
import Legend from "./Legend";
import AlertLevelLegend from "./alertLevelLegend";
import { searchResultSelector } from "../../../store/Homepage/homepageSlice";

const accessToken = process.env.REACT_APP_MAPBOX_ACCESS_TOKEN;

const tileURL = accessToken
  ? `https://api.mapbox.com/styles/v1/eric-ovio/ckesyzu658klw19s6zc0adlgp/tiles/{z}/{x}/{y}@2x?access_token=${accessToken}`
  : "https://server.arcgisonline.com/ArcGIS/rest/services/World_Imagery/MapServer/tile/{z}/{y}/{x}@2x";

const attribution = accessToken
  ? '© <a href="https://www.mapbox.com/about/maps/">Mapbox</a> © <a href="http://www.openstreetmap.org/copyright">OpenStreetMap</a> <strong><a href="https://www.mapbox.com/map-feedback/" target="_blank">Improve this map</a></strong>'
  : "";

const currentLocationMarker = L.divIcon({
  className: "current-position-marker",
  iconSize: L.point(16, 16, true),
});

const HomepageMap = ({
  initialCenter,
  initialZoom,
  classes,
}: HomepageMapProps) => {
  const [legendName, setLegendName] = useState<string>("");
  const [currentLocation, setCurrentLocation] = useState<[number, number]>();
  const [currentLocationAccuracy, setCurrentLocationAccuracy] = useState<
    number
  >();
  const [
    currentLocationErrorMessage,
    setCurrentLocationErrorMessage,
  ] = useState<string>();
  const loading = useSelector(reefsListLoadingSelector);
  const searchResult = useSelector(searchResultSelector);
  const ref = useRef<Map>(null);

  const onLocationSearch = () => {
    if (navigator.geolocation) {
      navigator.geolocation.getCurrentPosition(
        (position) => {
          const latLng = [
            position.coords.latitude,
            position.coords.longitude,
          ] as [number, number];
          setCurrentLocation(latLng);
          setCurrentLocationAccuracy(position.coords.accuracy);

          // zoom to user location
          const { current } = ref;
          if (current && current.leafletElement) {
            const map = current.leafletElement;
            const newZoom = Math.max(map.getZoom() || 6, 8);
            map.flyTo(latLng, newZoom, { duration: 2 });
          }
        },
        () => {
          setCurrentLocationErrorMessage("Unable to find your location");
        }
      );
    } else {
      setCurrentLocationErrorMessage(
        "Geolocation is not supported by your browser"
      );
    }
  };

  const onLocationErrorAlertClose = () =>
    setCurrentLocationErrorMessage(undefined);

  useEffect(() => {
    const { current } = ref;
    if (current && current.leafletElement) {
      const map = current.leafletElement;
      if (searchResult) {
        map.fitBounds([
          searchResult.bbox.southWest,
          searchResult.bbox.northEast,
        ]);
      }
    }
  }, [searchResult]);

  const onBaseLayerChange = ({ name }: LayersControlEvent) => {
    setLegendName(name);
  };

  return loading ? (
    <div className={classes.loading}>
      <CircularProgress size="4rem" thickness={1} />
    </div>
  ) : (
    <Map
      ref={ref}
      preferCanvas
      maxBoundsViscosity={1.0}
      className={classes.map}
      center={initialCenter}
      zoom={initialZoom}
      minZoom={2}
      worldCopyJump
      onbaselayerchange={onBaseLayerChange}
    >
      <Snackbar
        open={Boolean(currentLocationErrorMessage)}
        autoHideDuration={5000}
        anchorOrigin={{ vertical: "top", horizontal: "right" }}
        onClose={onLocationErrorAlertClose}
      >
        <Alert severity="error" onClose={onLocationErrorAlertClose}>
          {currentLocationErrorMessage}
        </Alert>
      </Snackbar>
      <TileLayer attribution={attribution} url={tileURL} />
      <SofarLayers />
      <ReefMarkers />
      {currentLocation && (
        <Marker icon={currentLocationMarker} position={currentLocation} />
      )}
      {currentLocation && currentLocationAccuracy && (
        <Circle
          center={{ lat: currentLocation[0], lng: currentLocation[1] }}
          radius={currentLocationAccuracy}
        />
      )}
      <Legend legendName={legendName} />
      <AlertLevelLegend />
      <div className="mapbox-wordmark" />
      <div className={classes.locationIconButton}>
        <IconButton onClick={onLocationSearch}>
          <MyLocationIcon color="primary" />
        </IconButton>
      </div>
    </Map>
  );
};

const styles = () =>
  createStyles({
    map: {
      flex: 1,
    },
    loading: {
      height: "100%",
      width: "100%",
      display: "flex",
      justifyContent: "center",
      alignItems: "center",
    },
    locationIconButton: {
      cursor: "pointer",
      display: "flex",
      alignItems: "center",
      justifyContent: "center",
      position: "absolute",
      left: 0,
      top: 80,
      zIndex: 1000,
      height: 34,
      width: 34,
      border: "2px solid rgba(0,0,0,0.2)",
      borderRadius: 5,
      margin: "10px 0 0 10px",
      backgroundColor: "white",
      backgroundClip: "padding-box",
    },
  });

interface HomepageMapIncomingProps {
  initialCenter: LatLng;
  initialZoom: number;
}

type HomepageMapProps = WithStyles<typeof styles> & HomepageMapIncomingProps;

export default withStyles(styles)(HomepageMap);
