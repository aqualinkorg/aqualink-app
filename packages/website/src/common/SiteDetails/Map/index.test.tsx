/* eslint-disable no-underscore-dangle */
import React from 'react';
import { Provider } from 'react-redux';
import { render } from '@testing-library/react';
import configureStore from 'redux-mock-store';

import { mockUser } from 'mocks/mockUser';
import Map from '.';

vi.mock('react-leaflet');

const mockStore = configureStore([]);

describe('Site Map', () => {
  let element: HTMLElement;
  beforeEach(() => {
    const store = mockStore({
      user: {
        userInfo: mockUser,
      },
      selectedSite: {
        draft: null,
      },
    });

    store.dispatch = vi.fn();

    element = render(
      <Provider store={store}>
        <Map
          polygon={{
            type: 'Polygon',
            coordinates: [[[0, 0]]],
          }}
          siteId={1}
          surveyPoints={[]}
        />
      </Provider>,
    ).container;
  });

  it('should render with given state from Redux store', () => {
    expect(element).toMatchSnapshot();
  });
});
