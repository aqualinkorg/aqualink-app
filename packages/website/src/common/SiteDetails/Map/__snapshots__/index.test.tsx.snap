// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Reef Map should render with given state from Redux store 1`] = `<div />`;
